include/deal.II/matrix_free/assemble_operator.h

```diff
@@ -0,0 +1,278 @@
+     * @param <tt>cache</tt> an already created CacheGraph object storing intermediate results.
+     * This allows to reuse the graph in case the sparsity pattern does not
+     change.
+     * If cache is empty (nullptr), it will be created.
+     */
+    template <typename Matrix,
+              typename Operator,
+              typename Pattern,
+              typename DomainVector,
+              typename RangeVector>
+    void
+    assemble_operator(
+      Matrix &                     output,
+      const Operator &             op,
+      const Pattern &              pattern,
+      const DomainVector &         dummy_domain,
+      const RangeVector &          dummy_range,
+      std::shared_ptr<GraphCache> &cache = std::shared_ptr<GraphCache>())
+    {
+      // Early exit for empty pattern.
+      if (pattern.n_nonzero_elements() == 0)
+        return;
+
+      // If not cache is given, initialize a new object.
+      if (!cache)
+        {
+          cache = std::make_shared<GraphCache>();
+
+          // Create a set with all the edges defined by the sparsity pattern.
+          // An edge E = (c1,c2) is contained iff there exists a row r,
+          // such that (r,c1) and (r,c2) are both contained in the pattern.
+          cache->make_graph(pattern);
+          cache->make_coloring();
+        }
+
+      // Coloring: for each edge (v1,v2) the colors of the vertices is
+      // different, i.e. c(v1) != c(v2). To obtain the matrix from the operator,
+      // we need to test (multiply) with one vector e per color. For each color
+      // c, e[i] = 1 if c(v[i]) == c else 0 (entries = {i : v[i] == c}). This
+      // gives us a vector col = sum_i A e_i = sum_i column_i(A). We need to
+      // decompose col into its contributions from the different columns. This
+      // is possible, as all the columns listed in entries have no non-zero row
+      // in common.
+      for (vertices_size_type color = 0; color < cache->num_colors; color++)
+        {
+          // Columns of color c.
+          std::vector<vertices_size_type> entries;
+          entries.reserve(pattern.max_entries_per_row());
+          for (vertices_size_type i = 0; i < pattern.n_cols(); i++)
+            if (cache->color_vec[i] == color)
+              entries.push_back(i);
+
+          // Construct the test-vector e = sum e_i for i in entries.
+          DomainVector e;
+          e.reinit(dummy_domain, false);
+          for (auto i : entries)
+            e[i] = 1.0;
+
+          // Apply to operator. image = sum A*e_i for i in entries.
+          RangeVector image;
+          image.reinit(dummy_range, false);
+          op.vmult(image, e);
+
+          // Split the result into its contributions from the single unit
+          // vectors, i.e. A*e_i. This is possible, since for each row, only one
+          // column in entry can store a value.
+          for (vertices_size_type r = 0; r < pattern.n_rows(); r++)
+            {
+              if (image[r] != 0.0)
```

Comment:
Cannot you pre-compute the indices based on the color?

Reply:
I was not sure if I should add that, as it requires to store n_colors * n_rows integers.